lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,174 @@
+# SUSE's openQA tests
+#
+# Copyright SUSE LLC
+# SPDX-License-Identifier: FSFAP
+# Maintainer: QE-SAP <qe-sap@suse.de>
+
+package sles4sap::console_redirection;
+use strict;
+use warnings;
+use testapi;
+use Carp qw(croak);
+use Exporter qw(import);
+use serial_terminal qw(select_serial_terminal);
+use Regexp::Common qw(net);
+
+=head1 SYNOPSIS
+
+Library that enables console redirection and file transfers from worker based VM to another host.
+Can be used for cases where worker VM is not the target host for API calls and command execution, but serves only as a jumphost.
+
+=cut
+
+our @EXPORT = qw(
+  connect_target_to_serial
+  disconnect_target_from_serial
+  verify_ip
+  redirection_init
+  check_serial_redirection
+);
+
+my $ssh_opt = '-o StrictHostKeyChecking=no -o ServerAliveInterval=60 -o ServerAliveCountMax=120';
+
+=head2 handle_login_prompt
+
+    handle_login_prompt();
+
+B<ssh_user> Login user
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub handle_login_prompt {
+    my $pwd = get_var('_SECRET_SUT_PASSWORD', $testapi::password);
+    set_serial_term_prompt();
+    my $serial_response = wait_serial(qr/Password:\s*$|:~/i);
+    # Handle password prompt if it appears
+    if (grep /Password:\s*$/, $serial_response) {
+        type_password $pwd;
+        send_key 'ret';
+    }
+    croak 'Failed to setup serial console' unless grep /:~/, $serial_response;
+}
+
+=head2 redirection_init
+
+    redirection_init();
+
+Do preparation before redirecting console. Gets base VM id and initial setup.
+Needs to be done only once.
+
+=cut
+
+sub redirection_init {
+    # This should get worker VM id before any redirection happening
+    # ID serves as identification of the 'base' VM to return to.
+    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+}
+
+=head2 set_serial_term_prompt
+
+    set_serial_term_prompt();
+
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub set_serial_term_prompt {
+    $testapi::distri->{serial_term_prompt} = '';    # reset previous prompt first
+    my $current_user = script_output('whoami');
+    croak 'SSH user not defined and/or "whoami" script failed.' unless defined($current_user);
+    $testapi::distri->{serial_term_prompt} = ($current_user eq 'root' ? '# ' : '> ');
+}
+
+=head2 connect_target_to_serial
+
+    connect_target_to_serial([ssh_user=>ssh_user, target_ip=>$target_ip]);
+
+B<ssh_user> Login user - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_USER
+B<target_ip> Target host IP - can be alternatively defined by OpenQA parameter REDIRECT_TARGET_IP
+Detects if login prompt appears and types the password.
+
+=cut
+
+sub connect_target_to_serial {
+    my (%args) = @_;
+    my $redirect_ip = $args{'target_ip'} // get_required_var('REDIRECT_TARGET_IP');
+    my $ssh_user = $args{'ssh_user'} // get_required_var('REDIRECT_TARGET_USER');
+
+    croak 'Missing "ssh_user" argument' unless $ssh_user;
```

Comment:
maybe these 2 `croak` are not needed due to the fact that `get_required_var` is used

Reply:
removed